driver: sensor: Add support for liteon ltr390 light and uvi module.

luke-holt opened the pull request:
Add support for liteon ltr390 ambient light and ultraviolet index sensor.

Datasheet:
https://optoelectronics.liteon.com/upload/download/DS86-2015-0004/LTR-390UV_Final_%20DS_V1%201.pdf

Testing environment:
Nordic nRF52840 DK
STM Nucleo F411RE
Adafruit 4831: https://www.adafruit.com/product/4831

Signed-off-by: Luke Holt <luke.holt55@gmail.com>

luke-holt committed:
driver: sensor: Add support for liteon ltr390 light and uvi module

Add support for liteon ltr390 ambient light and uv index sensor.

Added I2C driver and sample.

Tested on nrf52840dk and nucleo f411re with adafruit 4831.

Signed-off-by: Luke Holt <luke.holt55@gmail.com>

teburd commented on samples/sensor/ltr390/src/main.c:
```diff
@@ -0,0 +1,143 @@
+}
+#endif /* CONFIG_LTR390_TRIGGER */
+
+void main(void)
```
The project has been trying hard to avoid each sensor having its own sample but instead having a common sample application covering each sensor type.

It would be great if this sample could be reworked to do just that for light_polling sample akin to the accel_polling and magn_polling samples in the tree today.

It would probably be ok for the sensor to indicated it has specific features if those are needed in the sample, e.g. having a threshold attribute.

teburd commented on samples/sensor/ltr390/src/main.c:
```diff
@@ -0,0 +1,143 @@
+}
+#endif /* CONFIG_LTR390_TRIGGER */
+
+void main(void)
```
The project has been trying hard to avoid each sensor having its own sample but instead having a common sample application covering each sensor type.

It would be great if this sample could be reworked to do just that for light_polling sample akin to the accel_polling and magn_polling samples in the tree today.

It would probably be ok for the sensor to indicated it has specific features if those are needed in the sample, e.g. having a threshold attribute.

luke-holt replied:
Should I include the UVI channel in the light_polling sample? The accel_polling and magn_polling samples target `ACCEL_(X|Y|Z)` and `MAGN(X|Y|Z)` respectively.

If not, should I create a uvi_polling sample?

teburd replied:
That makes sense to me, @MaureenHelm any thoughts?

luke-holt replied:
I added a light_polling sample.

I left the ltr390 sample, because otherwise, there would be no demonstration of the trigger functionality. Should this instead be added to light_polling?

Any thoughts?

teburd commented on include/zephyr/drivers/sensor.h:
```diff
@@ -177,6 +177,9 @@ enum sensor_channel {
 	SENSOR_CHAN_GAUGE_DESIRED_CHARGING_CURRENT,
 
+	/** Ultraviolet Index (UVI) */
+	SENSOR_CHAN_UVI,
```
Neat that this exists! I knew of UV intensity sensors for chromatography but not of UVI sensors.

teburd commented on drivers/sensor/ltr390/ltr390_trigger.c:
```diff
@@ -0,0 +1,351 @@
+	void *als_raw)
+{
+	uint8_t *bytes = (uint8_t *)als_raw;
+	double l;
```
This should be done with integer math rather than float math, take a look around the tree at other sensor drivers to get a sense of how you might go about that

luke-holt replied:
Done.

teburd reviewed (changes requested):
Thank you for contributing this, please take a look at the feedback

luke-holt committed:
drivers: sensor: Rework conversions to use integer math

Reworked LTR390 driver to use integer math rather than floating point
math when converting from sensor_value structs to sensor-readable value.

Tested changes on nrf52840dk_nrf52480 and nucleo_f411re with ltr390
(Adafruit 4831).

Signed-off-by: Luke Holt <luke.holt55@gmail.com>

luke-holt committed:
samples: sensor: Added light_polling sample

Added light_polling sample that periodically samples ambient
light values in lux.

Signed-off-by: Luke Holt <luke.holt55@gmail.com>

luke-holt committed:
Merge branch 'main' into add_ltr390_sensor_driver

thedevleon commented:
Any reason this never got merged? Is there anything left to address? Would be happy to finish the work and get this merged.
